File: one_time_password_examples/token.rb
require 'rubygems'
require 'base64'
require 'openssl'
require 'typhoeus'
 
encodedkey = "RmFzdGx5IFRva2VuIFRlc3Q="
interval = 60
 
key = Base64.decode64(encodedkey)
 
number = Time.now.utc.to_i/interval
l = (number & 0xffffffff00000000) >>32
r = number & 0x00000000ffffffff

number = [r,l].pack('VV') 
 
digest = OpenSSL::HMAC.digest('sha256', key, number)
 
token = Base64.encode64(digest).strip()
 
response = Typhoeus::Request.get("http://token.fastly.com/token")
validation = Typhoeus::Request.get("http://token.fastly.com?#{token}")
 
p "Your Token:   #{token}"
p "Fastly Token: #{response.body}"
p "Validation:   #{validation.body}"

File: one_time_password_examples/token.rs
#[macro_use]
extern crate structure;

use std::time::{SystemTime};
use sha2::Sha256;
use hmac::{Hmac, Mac};


fn main() {
    type HmacSha256 = Hmac<Sha256>;

    let encodedkey = "RmFzdGx5IFRva2VuIFRlc3Q=";
    let interval = 60;

    let key = base64::decode(encodedkey).unwrap();
    let numstrct = structure::structure!("<Q");
    let number = numstrct.pack(SystemTime::now().duration_since(SystemTime::UNIX_EPOCH).unwrap().as_secs() / interval).unwrap();
    
    let mut hmac = HmacSha256::new_varkey(&key).unwrap();
    hmac.input(&number);
    let result = base64::encode(hmac.result().code());

    println!("Your token: {}", result);
    println!("Fastly token: {}", ureq::get("http://token.fastly.com/token").call().into_string().unwrap());
    println!("Validation: {}", ureq::get(&format!("http://token.fastly.com?{}", result)).call().into_string().unwrap());
}

